File: algorithms/WordBreakI/WordBreakI.py
class Solution(object):
    def wordBreak(self, s, wordDict):        
        wordSet = set(wordDict)
        n = len(s)
        dp = [False] * n
        
        for i in range(n):
            word = s[:i + 1]
            
            if word in wordSet:
                dp[i] = True
                continue
            
            for j in range(i):
                right = s[j + 1:i + 1]
                if dp[j] and right in wordDict:
                    dp[i] = True
                    break
        
        return dp[n - 1]

class Solution(object):
    def wordBreakHelper(self, s, wordSet, start, cache):
        if start == len(s):
            return True
        
        if start in cache:
            return cache[start]
        
        for end in range(start + 1, len(s) + 1):
            if s[start:end] in wordSet and self.wordBreakHelper(s, wordSet, end, cache):
                cache[start] = True 
                return True
         
        cache[start] = False
        return False
            
        
    def wordBreak(self, s, wordDict):
        return self.wordBreakHelper(s, set(wordDict), 0, {})

File: algorithms/WordBreakI/WordBreakI.cpp
#include <iostream>
#include <unordered_set>
using namespace std;

class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> wordSet(wordDict.begin(), wordDict.end());
        int n = s.size();

        // dp[i] represents that substring from 0 to i can be broken or not
        bool dp[n] = {false};

        for(int i = 0; i < n; i++) {
            string word = s.substr(0, i + 1);

            // Check if the substring from 0 to i is in dict or not
            if(wordSet.count(word) > 0) {
                dp[i] = true;
                continue;
            }

            // If not, break the substring and check
            for(int j = 0 ; j < i; j++) {
                if(dp[j]) {
                    string right = s.substr(j + 1, i - j);
                    if(wordSet.count(right) > 0) {
                        dp[i] = true;
                        break;
                    }
                }
            }
        }
        return dp[n - 1];
    }
};
